### PR TITLE
Ingestion improvements

### PR DESCRIPTION
**Pull request by nullstyle**
This PR makes a number of changes to ingestion (and other tangential systems) in an effort to improve reliability across the various stellar-core db states that can occur.

Most significantly, this PR changes the "empty history database" behavior that horizon exhibits.  Prior to this PR, horizon would try to find a suitable ledger in the past from which to start ingestion from.  After this PR, horizon will start ingesting from the latest ledger in stellar-core.  This much simpler behavior is more robust in the face of a stellar-core sql database that may contain one or more gaps in history.

Additionally, we introduce history backfill support through the `horizon db backfill` command.  This command allows an administrator to backfill the history database, starting at horizon's elder ledger.  An example command:  `horizon db backfill 20` would ingest 20 ledgers of history backwards from the earliest ledger in horizon's database.

Finally, this PR changes horizon to have it ingest and store the ledger header xdr structs in the `history_ledgers` table.  This will allow us to re-ingest ledger data in the future without the need to keep the history in stellar-core.

## Other Random Changes

- Update the version of go we use to build our release artifacts to go 1.9
- A simple dockerfile was added that builds the monorepo tools/services

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,11 @@
+FROM golang:1.9
+RUN bash -c "curl https://glide.sh/get | sh"
+WORKDIR /go/src/github.com/stellar/go
+
+COPY glide.lock /go/src/github.com/stellar/go
```
Comment (nikhilsaraf) [inline Dockerfile]: is this file meant to be in the `stellar/go` repo? -- same for `.dockerignore` above

File: services/horizon/internal/docs/reference/resources/ledger.md [hunk 2]
```diff
@@ -65,7 +67,9 @@ To learn more about the concept of ledgers in the Stellar network, take a look a
   "fee_pool": "0.0025600",
   "base_fee": 100,
   "base_reserve": "10.0000000",
-  "max_tx_set_size": 50
+  "max_tx_set_size": 50,
+  "protocol_version": 8,
+  "header_xdr": "...",
```
Comment (nikhilsaraf) [inline services/horizon/internal/docs/reference/resources/ledger.md]: is it supposed to be `...` here or is this just a placeholder?

File: services/horizon/internal/ingest/cursor.go [hunk 3]
```diff
@@ -113,6 +105,28 @@ func (c *Cursor) NextLedger() bool {
...
 }
 
+func (c *Cursor) incrementLg() bool {
+	isReverse := c.FirstLedger > c.LastLedger
+
+	if c.lg == 0 {
+		c.lg = c.FirstLedger
+	} else {
+		increment := int32(1)
+		if isReverse {
```
Comment (nikhilsaraf) [inline services/horizon/internal/ingest/cursor.go]: suggestion: could do 3 top-level if conditions: `c.lg == 0`, `isReverse`, `else` to avoid this inner condition --we are already operating on c.lg in the first one.

File: services/horizon/internal/ingest/cursor.go [hunk 4]
```diff
@@ -113,6 +105,28 @@ func (c *Cursor) NextLedger() bool {
...
+			increment = int32(-1)
+		}
+		c.lg += increment
+	}
+
+	if (!isReverse && c.lg > c.LastLedger) || (isReverse && c.lg < c.LastLedger) {
```
Comment (nikhilsaraf) [inline services/horizon/internal/ingest/cursor.go]: can we extract these two sub-conditions into variables so it's easier to reason about?

not sure whether `c.lg` is supposed to be = to `c.LastLedger` when iteration is completed but want to pose that question to you to avoid a potential off-by-one error here.

File: services/horizon/internal/ingest/cursor.go [hunk 5]
```diff
@@ -113,6 +105,28 @@ func (c *Cursor) NextLedger() bool {
 	return true
 }
 
+func (c *Cursor) incrementLg() bool {
+	isReverse := c.FirstLedger > c.LastLedger
```
Comment (nikhilsaraf) [inline services/horizon/internal/ingest/cursor.go]: this is fine but wanted to suggest having an `iterator` or `counter` type which handles all this directional logic. then `c.lg` can be of that type, separating out the business logic from the functional logic.

File: services/horizon/internal/ingest/cursor_test.go [hunk 6]
```diff
@@ -56,4 +56,23 @@ func TestCursor(t *testing.T) {
...
+	tt.Require.True(c.NextLedger())
+	tt.Require.Equal(uint32(10), c.Ledger().Sequence)
+	tt.Require.True(c.NextLedger())
+	tt.Require.Equal(uint32(9), c.Ledger().Sequence)
+	tt.Require.True(c.NextLedger())
+	tt.Require.Equal(uint32(8), c.Ledger().Sequence)
+	tt.Require.True(c.NextLedger())
+	tt.Require.Equal(uint32(7), c.Ledger().Sequence)
+
+	tt.Require.False(c.NextLedger())
```
Comment (nikhilsaraf) [inline services/horizon/internal/ingest/cursor_test.go]: what is the value of `c.Ledger().Sequence` at this point? `8`, or `7`?
could add an assertion for that to lock that behavior down via this test

File: services/horizon/internal/ingest/system.go [hunk 7]
```diff
@@ -269,33 +244,25 @@ func (i *System) runOnce() {
 		return
 	}
 
-	cursor, err := i.newCursor()
-	if err != nil {
-		log.Errorf("failed to create ingestion cursor: %s", err)
+	if ls.CoreLatest == 1 {
```
Comment (nikhilsaraf) [inline services/horizon/internal/ingest/system.go]: do we need to ping stellar-core here to see if its status is syncd?
stellar could be in the process of catching up from a gap (ls.CoreLatest > 0 but it's still syncing) --not sure if this line was supposed to be the fix for that but wanted to check that with you.

File: services/horizon/internal/ingest/system.go [hunk 8]
```diff
@@ -269,33 +244,25 @@ func (i *System) runOnce() {
...
 
 	// 2.
 	if ls.HistoryLatest == 0 {
 		log.Infof(
-			"history db is empty, starting ingestion from ledger %d",
-			is.Cursor.FirstLedger,
+			"history db is empty, establishing base at ledger %d",
+			ls.CoreLatest,
 		)
+		is.Cursor = NewCursor(ls.CoreLatest, ls.CoreLatest, i)
```
Comment (nikhilsaraf) [inline services/horizon/internal/ingest/system.go]: @nullstyle, was expecting this to start from `ls.CoreLatest - HISTORY_STALE_THRESHOLD` till `ls.CoreLatest` so we can always have `HISTORY_STALE_THRESHOLD` number of ledgers even for a fresh db.

File: services/horizon/internal/test/scenarios/account_merge-core.sql [hunk 9]
```diff
@@ -379,11 +379,12 @@ INSERT INTO scpquorums VALUES ('d9d851d5b6955c86634bcbcfed1337f717ca03dd9228a92a
...
 
 INSERT INTO storestate VALUES ('databaseschema                  ', '5');
+INSERT INTO storestate VALUES ('networkpassphrase               ', 'Test SDF Network ; September 2015');
 INSERT INTO storestate VALUES ('forcescponnextlaunch            ', 'false');
-INSERT INTO storestate VALUES ('lastclosedledger                ', '3353357fc019e2b58618e6abce8664fbac5ce0c87ef4f573e3c51be2fab5087e');
+INSERT INTO storestate VALUES ('lastclosedledger                ', 'd31840d2ab69b30accae9017be46afccb981b06342262b8214ac2ba70deb6f1f');
 INSERT INTO storestate VALUES ('historyarchivestate             ', '{
     "version": 1,
-    "server": "v0.6.3-66-gd66575cf",
+    "server": "v9.0.0-4-g59482f9d",
```
Comment (nikhilsaraf) [inline services/horizon/internal/test/scenarios/account_merge-core.sql]: wanted to bring this to your attention, is this supposed to have jumped from `v0.6..` to `v9.0..`?

File: services/horizon/internal/docs/reference/resources/ledger.md (hunk 2 above)
Comment (nullstyle) [reply services/horizon/internal/docs/reference/resources/ledger.md]: it's a placeholder.  these are just the docs.

File: services/horizon/internal/ingest/cursor.go (hunk 5 above)
Comment (nullstyle) [reply services/horizon/internal/ingest/cursor.go]: YAGNI

File: services/horizon/internal/ingest/cursor.go (hunk 3 above)
Comment (nullstyle) [reply services/horizon/internal/ingest/cursor.go]: They are logically separate if statements.  I prefer them separate because the likely refactoring would be to pull the increment functionality into its own func.

File: services/horizon/internal/ingest/cursor.go (hunk 4 above)
Comment (nullstyle) [reply services/horizon/internal/ingest/cursor.go]: I'm happy with the if statement.  I'll add a comment to describe the conditional in simpler terms.

The reseting of a cursor back to its zero-value after a complete traversal is by design.

File: services/horizon/internal/ingest/cursor_test.go (hunk 6 above)
Comment (nullstyle) [reply services/horizon/internal/ingest/cursor_test.go]: `c.Ledger()` panics on a nil error when you're not in the midst of an active iteration.

File: services/horizon/internal/ingest/system.go (hunk 7 above)
Comment (nullstyle) [reply services/horizon/internal/ingest/system.go]: We can't depend upon stellar-core's sync status in most cases.  It's not stable or correct enough to reliably program against.  For example, it reports "Synced!" on bootup even prior to joining SCP.

This method of detecting a newly booted stellar core is much more stable.

File: services/horizon/internal/ingest/system.go (hunk 8 above)
Comment (nullstyle) [reply services/horizon/internal/ingest/system.go]: This is an intentional design change to improve the stability of ingestion.  Admins may backfill if they desire to load history from the past.

File: services/horizon/internal/test/scenarios/account_merge-core.sql (hunk 9 above)
Comment (nullstyle) [reply services/horizon/internal/test/scenarios/account_merge-core.sql]: its fine.

File: Dockerfile (hunk 1 above)
Comment (nullstyle) [reply Dockerfile]: Yes.  I've added a simple docker file to our monorepo that builds our tools.

File: services/horizon/internal/test/scenarios/account_merge-core.sql (hunk 9 above)
Comment (bartekn) [reply services/horizon/internal/test/scenarios/account_merge-core.sql]: @nikhilsaraf yes, now the `major` part of the stellar-core version is the protocol version [1]:
> To make things simpler to track and understand, we’re switching to version numbers setup as protocol_version.release_number.patch_number.

[1] https://github.com/stellar/stellar-core/releases/tag/v9.0.0

**Comment by nullstyle** [comment]
@nikhilsaraf this PR should be good to go now.  Please re-review :)

**Comment by nikhilsaraf** [comment]
deferring to bartek's approval, unless you want another pair of eyes on it.
